specs/core/0_beacon-chain.md

```diff
@@ -317,27 +285,130 @@ Unless otherwise indicated, code appearing in `this style` is to be interpreted
 }
 ```
 
-#### `ProposalSignedData`
+#### `ProposerSlashingRecord`
+
+```python
+{
+    # Proposer index
+    'proposer_index': 'uint24',
+    # First proposal data
+    'proposal_data_1': ProposalSignedData,
+    # First proposal signature
+    'proposal_signature_1': '[uint384]',
+    # Second proposal data
+    'proposal_data_2': ProposalSignedData,
+    # Second proposal signature
+    'proposal_signature_2': '[uint384]',
+}
+```
+
+#### `CasperSlashingRecord`
+
+```python
+{
+    # First vote
+    vote_1: SpecialAttestationData,
+    # Second vote
+    vote_2: SpecialAttestationData,
+}
+```
+
+#### `SpecialAttestationData`
+
+ ```python
+{
+    # Proof-of-custody indices (0 bits)
+    'aggregate_signature_poc_0_indices': '[uint24]',
+    # Proof-of-custody indices (1 bits)
+    'aggregate_signature_poc_1_indices': '[uint24]',
+    # Attestation data
+    'data': AttestationData,
+    # Aggregate signature
+    'aggregate_signature': '[uint384]',
+}
+```
+
+#### `DepositRecord`
+
+```python
+{
+    # Receipt Merkle branch
+    'merkle_branch': '[hash32]',
+    # Merkle tree index
+    'merkle_tree_index': 'uint64',
+    # Deposit data
+    'deposit_data': {
+        # Deposit parameters
+        'deposit_parameters': DepositParametersRecord,
+        # Value in Gwei
+        'value': 'uint64',
+        # Timestamp from deposit contract
+        'timestamp': 'uint64',
+    },
+}
+```
+
+#### `ExitRecord`
+
+```python
+{
+    # Minimum slot for processing exit
+    'slot': 'unit64',
+    # Index of the exiting validator
+    'validator_index': 'uint64',
+    # Validator signature
+    'signature': '[uint384]',
+}
+```
+
+### Beacon chain blocks
+
+#### `BeaconBlockHeader`
 
 ```python
 {
-    # Slot number
     'slot': 'uint64',
-    # Shard number (`BEACON_CHAIN_SHARD_NUMBER` for beacon chain)
-    'shard': 'uint64',
-    # Block hash
-    'block_hash': 'hash32',
+    # Skip list of ancestor beacon block hashes
+    # i'th item is the most recent ancestor whose slot is a multiple of 2**i for i = 0, ..., 31
+    'ancestor_hashes': ['hash32'],
+    'state_root': 'hash32',
+    'randao_reveal': 'hash32',
+    'candidate_pow_receipt_root': 'hash32',
+    'signature': ['uint384'],
 }
 ```
 
-#### `SpecialRecord`
+#### `BeaconBlockBody`
 
 ```python
 {
-    # Kind
-    'kind': 'uint64',
-    # Data
-    'data': 'bytes',
+    'attestations': [AttestationRecord],
+    'proposer_slashings': [ProposerSlashingRecord],
+    'casper_slashings': [CasperSlashingRecord],
+    'deposits': [DepositRecord],
+    'exits': [ExitRecord],
+}
+```
+
+#### `BeaconBlock`
+
+```python
+{
+    'header': BeaconBlockHeader,
+    'body': BeaconBlockBody,
+}
```

Comment:
Not sure I like this; I'd feel better with `body` being an element of the header.

Remember that at the SSZ tree hashing level, that would just mean that the body gets represented by a nice clean 32 byte hash.

Reply:
Are you suggesting the following? Just doing away with `BeaconBlockHeader`? I like that.

```
#### BeaconBlock

{
    'slot': 'uint64',
    # Skip list of ancestor beacon block hashes
    # i'th item is the most recent ancestor whose slot is a multiple of 2**i for i = 0, ..., 31
    'ancestor_hashes': ['hash32'],
    'state_root': 'hash32',
    'randao_reveal': 'hash32',
    'candidate_pow_receipt_root': 'hash32',
    'signature': ['uint384'],
    'body': BeaconBlockBody,
}